PM Recipient Network Integration

eladmallel opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Makes use of the PM Recipient side in the Livepeer networking, allowing Bs to receive Os' payment parameters, and for O to process incoming PM tickets from B.

**Specific updates (required)**

- O responds to `GetOrchestratorInfo` with `TicketParams` which B must use to send valid PM tickets
- O processes incoming PM tickets, starting with parsing them from the HTTP header and passing them down to `Recipient.ReceiveTicket`
- Some refactoring in support of this change

**How did you test each of these updates (required)**

- Added unit tests for the new code and no previous tests have regressed


**Does this pull request close any open issues?**

- Fixes https://github.com/livepeer/go-livepeer/issues/635


**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

eladmallel committed:
PM: O processes payments attached to segments

eladmallel committed:
PM: new mockOrchestrator

Because I want the ability to define the behavior of more methods
that are already implemented in the stub type.

eladmallel committed:
PM: recipient TicketParams refactor

Seems like it was returning an error only because of the specific
implementation for random bytes generation. It seems sensible to
refactor that away so that upon integration with RPC code, we don't
have to account for this error that should never really happen anyway.

eladmallel committed:
PM: ticket params in OrchestratorInfo response

yondonfu commented on pm/recipient.go:
```diff
@@ -24,7 +24,7 @@ type Recipient interface {
 
 	// TicketParams returns the recipient's currently accepted ticket parameters
 	// for a provided sender ETH adddress
-	TicketParams(sender ethcommon.Address) (*TicketParams, error)
```
🎉

yondonfu commented on core/orchestrator.go:
```diff
@@ -79,6 +80,40 @@ func (orch *orchestrator) TranscoderResults(tcId int64, res *RemoteTranscoderRes
 	orch.node.transcoderResults(tcId, res)
 }
 
+func (orch *orchestrator) ProcessPayment(payment net.Payment, manifestID ManifestID) error {
```
I think we'll get a nil pointer exception on L94 if the node is running in off-chain mode since `orch.node.Recipient` will be nil. The rest of the orchestrator methods that depend on the presence of a ETH client first check that both the node and the ETH client are not nil. Perhaps we can add a similar check here:

```
if orch.node == nil || orch.node.Recipient == nil {
    return nil
}
```

eladmallel replied:
Awesome tip, thank you! I had that thought for a moment but then felt like it might be too defensive, forgetting all about off-chain mode! My bad!

yondonfu commented on core/orchestrator.go:
```diff
@@ -79,6 +80,40 @@ func (orch *orchestrator) TranscoderResults(tcId int64, res *RemoteTranscoderRes
+	return nil
+}
+
+func (orch *orchestrator) TicketParams(sender ethcommon.Address) *net.TicketParams {
```
Echoing my comment for `ProcessPayment` - I think we'll get a nil pointer exception on L107 if the node is running in off-chain mode since `orch.node.Recipient` will be nil. The rest of the orchestrator methods that depend on the presence of a ETH client first check that both the node and the ETH client are not nil. Perhaps we can add a similar check here:

```
if orch.node == nil || orch.node.Recipient == nil {
    return nil
}
```

eladmallel replied:
Thank you, fixed!

yondonfu commented on server/rpc_test.go:
```diff
@@ -205,3 +222,207 @@ func TestPing(t *testing.T) {
+
+func (o *mockOrchestrator) ServiceURI() *url.URL {
+	args := o.Called()
+	return args.Get(0).(*url.URL)
```
Might be worth being cautious with type asserting on nil here and checking for that in case this is ever used in the future to return a nil value

eladmallel replied:
haha I was being lazy here thinking if someone ran into that in the future they will notice (tests will panic) and they will fix it :) but since you mind it, I will add the quick fix :)

eladmallel committed:
PM: fixed nil pointer bug + test cleanup

eladmallel committed:
PM: minor cleanup

Protecting against a nil case that isn't happening right now, but should
provide better UX for the next developer who'll want to simulate a nil
case.

j0sh commented on server/rpc.go:
```diff
@@ -242,6 +245,9 @@ func getOrchestrator(context context.Context, orch Orchestrator, req *net.Orches
 		tr.Storage = []*net.OSInfo{os.GetInfo()}
 	}
+
+	tr.TicketParams = orch.TicketParams(addr)
```
Since TicketParams is unconditionally assigned, can we add this where `tr` is defined?

j0sh commented on server/rpc.go:
```diff
@@ -264,6 +284,11 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
 
+	err = processPayment(orch, r.Header.Get(PaymentHeader), segData.ManifestID)
+	if err != nil {
+		glog.Errorf("Error processing payment: %v", err)
```
We continue to process the segment even if the payment fails?

eladmallel replied:
Thank you for bringing this up - I wasn't sure what's the best move here, specifically for MVS. Do you think we should just not process the segment if there's an error here?

j0sh replied:
I have to dig up the discussion(s?), but IIRC we had a few scenarios that would justify a "grace period" where we'd continue execution even if the payment fails, and a few other scenarios that justified a hard stop.

I believe we were able to whittle down the MVS scope such that we could avoid implementing a grace period, leaving us with just hard stops. If that's still true, then I guess the appropriate thing to do here would be to send an error response and exit early.

yondonfu replied:
I agree with @j0sh point about sending an error response and exiting if the node is in *on-chain* mode. That being said, could there be a problem if we exit early as is when the node is in *off-chain* mode when the broadcaster tries to unmarshal an empty buffer in `processPayment`? If so, then the orchestrator will always exit early. Perhaps there should also be a check in `processPayment` to return nil if the node is in off-chain mode

j0sh replied:
IIRC, the deserialized `payment` that we receive from protobufs will be empty in off-chain mode -- probufs won't return an error on missing fields since all fields are optional in protobufs v3. The `orch.ProcessPayment` method does incorporate an off-chain check, so we should be all set here.

Would be good to test in `-offchain` mode just to be sure, though.

yondonfu replied:
Ah nice. Might be worthwhile to add an extra test with an empty payment string header too

eladmallel replied:
Awesome feedback - changed the code accordingly!

j0sh commented on server/rpc_test.go:
```diff
@@ -205,3 +222,207 @@ func TestPing(t *testing.T) {
+}
+
+type mockOrchestrator struct {
+	mock.Mock
```
We already have a `stubOrchestrator`, could the `mock.Mock` be worked into that rather than re-defining a separate set of stubs?

eladmallel replied:
I believe it's possible, but since other tests are using `stubOrchestrator` I preferred to avoid that refactor for now, for the sake of getting us faster to the point of integration tests. When do you think is the best time to tackle this?

yondonfu replied:
@j0sh curious to hear your take on this related [discussion](https://github.com/livepeer/go-livepeer/pull/657#discussion_r244420843) about the distinction between stubs and mocks and the possibility of having both.

j0sh replied:
Naming conventions aside (which I also don't really have a preference for), I was mostly wondering if the Mock could be incorporated into `stubSegment` without affecting existing tests. If that can't be accomplished cleanly, then it makes sense to have separate mocked stubs.

Are we set on using mocks instead of manual stubs? I can see the helpers saving us some work (eg, counting method calls), but the practice of setting expectations in advance prior to execution is fairly different from the post-execution assertions that we use everywhere else. My initial feeling is that I'd prefer to have a uniform way of checking those conditions, but I'm open if the consensus is that the mixed approach saves us time/complexity/etc.

yondonfu replied:
> I can see the helpers saving us some work (eg, counting method calls)

For this specific example, the `testify/mock` package seems to have a [Times()](https://godoc.org/github.com/stretchr/testify/mock#Call.Times) method that can be used to set the number of times that a mocked method should return a value and then the test would assert that the method was called the expected number of times using [AssertExpectations()](https://godoc.org/github.com/stretchr/testify/mock#Mock.AssertExpectations). I *think* this should do the trick? I haven't tested this in practice yet though.

> My initial feeling is that I'd prefer to have a uniform way of checking those conditions, but I'm open if the consensus is that the mixed approach saves us time/complexity/etc.

I personally have found the ability to set expectations prior to execution to be useful and feels cleaner than defining custom state variables in stubs. For that reason, I feel ok about a mixed approach right now.

j0sh replied:
The ability to count invocations of a method without writing boilerplate for each occurrence is indeed nice, and I'd like it even better if that could be checked after the fact, eg `assert.Equals(1, mockObj.MethodCallCount("foo"))` . If `AssertExpectations` can give us that, all the better.

yondonfu replied:
Ah there is actually an [AssertNumberOfCalls()](https://godoc.org/github.com/stretchr/testify/mock#Mock.AssertNumberOfCalls) method which does the trick too:

```
m.AssertNumberOfCalls(t, "Bar", 3)
```

j0sh replied:
I suppose that works. Would be more consistent to just return the count somehow, which composes better with the existing `assert` functionality -- passing in `t` every time is ugly.

j0sh commented on pm/helpers.go:
```diff
@@ -22,42 +21,18 @@ func VerifySig(addr ethcommon.Address, msg, sig []byte) bool {
-
-	return ethcommon.BytesToHash(key[:])
+func RandHash() ethcommon.Hash {
+	return ethcommon.BytesToHash(RandBytes(32))
```
Beautiful set of cleanups.

eladmallel committed:
PM: cleanup

yondonfu commented on core/orch_test.go:
```diff
@@ -286,65 +285,53 @@ func TestGetSegmentChan(t *testing.T) {
 	recipient.On("ReceiveTicket", mock.Anything, mock.Anything, mock.Anything).Return("some sessionID", true, errors.New("mock error"))
 
-	err = orch.ProcessPayment(defaultPayment(t), ManifestID("some manifest"))
+	orch.ProcessPayment(defaultPayment(t), ManifestID("some manifest"))
```
Check error here?

eladmallel committed:
PM: test fix

eladmallel committed:
PM: test for no payment header

To make sure this implementation supports "offchain mode" where two
nodes wants to work together without any ETH involved.

eladmallel committed:
PM: O rejects segment without payment

Except for offchain mode in which case it should process the segment
without issue.

yondonfu commented on server/rpc.go:
```diff
@@ -286,6 +286,8 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
 	err = processPayment(orch, r.Header.Get(PaymentHeader), segData.ManifestID)
 	if err != nil {
 		glog.Errorf("Error processing payment: %v", err)
+		http.Error(w, err.Error(), http.StatusPaymentRequired)
```
❤️ that this is an HTTP status code

yondonfu reviewed (approved):
LGTM

eladmallel committed:
Merge branch 'master' into em/pm-recipient-network-integration
